HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,90 @@
+# Initialize Spark
+from pyspark import SparkContext
+sc = SparkContext()
+sc.setLogLevel("ERROR")
+
+# Helper functions for cleaning wiki data
+def link_string_to_KV(s):
+    src, dests = s.split(': ')
+    dests = [int(to) for to in dests.split(' ')]
+    return (int(src), dests)
+
+# Helper function for updating path
+def get_path(x):
+    newChars = x[1][1] 
+    oldPath = x[1][0][1]
+    return [(c, [len(oldPath + [c]), oldPath + [c]]) for c in newChars]
+
+# Finds the shortest path between two nodes
+def find_path(searchGraph, startNodeID, endNodeID):
+    
+    # Initialize key variables
+    searchNodes = sc.parallelize([startNodeID]).map(lambda x: (x, x))
+    explored = sc.parallelize([startNodeID]).map(lambda x: (x, x))
+    frontier = sc.parallelize([startNodeID]).map(lambda x: (x, [1, [x]]))
+    targetNodes = sc.parallelize([])
+    dist = 1
+    
+    # Keep looping until the target node has been found
+    while targetNodes.isEmpty():
+
+        dist += 1
+        
+        # Look up new children nodes
+        children = searchGraph.join(searchNodes).map(lambda x: (x[0], x[1][0])).partitionBy(256)
```

Comment:
That's interesting - did you get out-of-memory problems locally or when running on AWS?  I found that the large number of partitions ran very smoothly on both and did help.